pkg/authorization/api/types.go

```diff
@@ -434,3 +434,90 @@ type ClusterRoleList struct {
 	Items []ClusterRole
 }
+
+// RoleBindingRestriction is an object that can be matched against a subject
+// (user, group, or service account) to determine whether rolebindings on that
+// subject are allowed in the namespace to which the RoleBindingRestriction
+// belongs.  If any one of those RoleBindingRestriction objects matches
+// a subject, rolebindings on that subject in the namespace are allowed.
+type RoleBindingRestriction struct {
+	unversioned.TypeMeta
+
+	// Standard object's metadata.
+	kapi.ObjectMeta
+
+	// Spec defines the matcher.
+	Spec RoleBindingRestrictionSpec
+}
+
+// RoleBindingRestrictionSpec defines a rolebinding restriction.  Exactly one
+// field must be non-nil.
+type RoleBindingRestrictionSpec struct {
+	// UserRestriction matches against user subjects.
+	UserRestriction *UserRestriction
+
+	// GroupRestriction matches against group subjects.
+	GroupRestriction *GroupRestriction
+
+	// ServiceAccountRestriction matches against service-account subjects.
+	ServiceAccountRestriction *ServiceAccountRestriction
+}
+
+// RoleBindingRestrictionList is a collection of RoleBindingRestriction objects.
+type RoleBindingRestrictionList struct {
+	unversioned.TypeMeta
+
+	// Standard object's metadata.
+	unversioned.ListMeta
+
+	// Items is a list of RoleBindingRestriction objects.
+	Items []RoleBindingRestriction
+}
+
+// UserRestriction matches a user either by a string match on the user name,
+// a string name on the name of a group to which the user belongs, or a label
+// selector.
+type UserRestriction struct {
+	// Users specifies a list of literal user names.
+	Users []string
+
+	// Groups is a list of groups used to match against an individual user's
+	// groups. If the user is a member of one of the whitelisted groups, the user
+	// is allowed to be bound to a role.
+	Groups []string
+
+	// Selectors specifies a list of label selectors.
+	Selectors []*unversioned.LabelSelector
+}
+
+// GroupRestriction matches a group either by a string match on the group name
+// or a label selector.
+type GroupRestriction struct {
+	// Groups specifies a list of literal group names.
+	Groups []string
+
+	// Selectors specifies a list of label selectors.
+	Selectors []*unversioned.LabelSelector
+}
+
+// ServiceAccountRestriction matches a service account by a string match on
+// either the service-account name or the name of the service account's
+// namespace.
+type ServiceAccountRestriction struct {
+	// ServiceAccounts specifies a list of literal service-account names.
+	ServiceAccounts []*ServiceAccountReference
+
+	// Namespaces specifies a list of literal namespace names.  ServiceAccounts
+	// from inside the whitelisted namespaces are allowed to be bound to roles.
+	Namespaces []string
+}
+
+// ServiceAccountReference specifies a service account and namespace by their
+// names.
+type ServiceAccountReference struct {
+	// Name is the name of the service account.
+	Name string
+
+	// Namespace is the name of the namespace.
```

Comment:
If the namespace is empty, then the reference means "SA with this name in the current namespace"